io-classes/src/Control/Monad/Class/MonadMVar.hs

```diff
@@ -240,102 +273,134 @@ takeMVarDefault (MVar tv) = mask_ $ do
-        MVarEmpty blockedq -> do
-          wakevar <- newTVar Nothing
-          writeTVar tv (MVarEmpty (Deque.snoc wakevar blockedq))
-          return (Left wakevar)
-
-        -- if it's full we grab the value, and also complete the action of the
-        -- next thread blocked in putMVar, by setting the new MVar value and
-        -- unblocking them.
-        MVarFull x blockedq ->
-          case Deque.uncons blockedq of
-            Nothing ->
+        -- It's empty, add ourselves to the end of the 'take' blocked queue.
+        MVarEmpty takeq readq -> do
+          takevar <- newTVar Nothing
+          writeTVar tv (MVarEmpty (Deque.snoc takevar takeq) readq)
+          return (Left takevar)
+
+        -- It's full. If there's at least one thread blocked in putMVar, wake
+        -- up the first one leaving the MVar full with the new put value.
+        -- Otherwise the MVar becomes empty.
+        MVarFull x putq ->
+          case Deque.uncons putq of
+            Nothing -> do
+              writeTVar tv (MVarEmpty mempty mempty)
               return (Right x)
 
-            Just ((x', wakevar), blockedq') -> do
-              writeTVar wakevar True
-              writeTVar tv (MVarFull x' blockedq')
+            Just ((x', putvar), putq') -> do
+              writeTVar putvar True
+              writeTVar tv (MVarFull x' putq')
               return (Right x)
 
     case res of
-      -- we have to block on our own wakevar until we can complete the read
-      Left wakevar ->
-        atomically (readTVar wakevar >>= maybe retry return)
+      -- we have to block on our own takevar until we can complete the read
+      Left takevar ->
+        atomically (readTVar takevar >>= maybe retry return)
         `catch` \e@SomeAsyncException {} -> do
           atomically $ do
             s <- readTVar tv
             case s of
-              MVarEmpty blockedq -> do
+              MVarEmpty takeq readq -> do
                 -- async exception was thrown while were were blocked on
-                -- wakevar; we need to remove it from 'blockedq', otherwise we
+                -- takevar; we need to remove it from 'takeq', otherwise we
                 -- will have a space leak.
-                let blockedq' = Deque.filter (/= wakevar) blockedq
-                writeTVar tv (MVarEmpty blockedq')
-              -- the exception was thrown while we were blocked on 'wakevar', so
-              -- the 'MVar' must not be full.
-              MVarFull {} -> error "takeMVarDefault: invariant violation"
+                let takeq' = Deque.filter (/= takevar) takeq
+                writeTVar tv (MVarEmpty takeq' readq)
+
+              -- This case is unlikely but possible if another thread ran
+              -- first and modified the mvar. This situation is fine as far as
+              -- space leaks are concerned because it means our wait var is no
+              -- longer in the wait queue.
+              MVarFull {} -> return ()
```

Comment:
I'm imagining this scenario that can trigger this case:

* Start with an empty MVar
* Thread 0: performs `putMVar x`, puts value `x` inside
* Thread 1: performs `putMVar y`, is blocked on putting `y`
* Thread 1: an asynchronous exception is thrown
* Thread 2: performs `takeMVar`, gets value `x`,  puts value `y` inside because it is still in `takeq`
* Thread 1: handles the asynchronous exception, but sees its `takevar` has already been removed 

Now there is no space leak, but the next `takeMVar`/`readMVar` will still receive value `y` (depending on whether I'm seeing this right, of course). Is this problematic? Or even preventable?

Reply:
Looking at this scenario again, is the case unlikely because it relies on a `takeMVar` being performed in between throwing the exception and handling the exception?

Reply:
That scenario is also one I've worried about. It results in a "spurious update".

In the real MVar impl, I think what would happen there is that either the thread 1 `putMVar` succeeds (updating the MVar value) and thread 1 can continue running with async exceptions masked, or the thread 1 `putMVar` is interrupted with the exception and the MVar value does not change.

In this impl we can have another behaviour: the thread 1 `putMVar` succeeds (updating the MVar value) but the thread receives the async exception despite having async exceptions masked. That's because it's always ok to get an async exception when the thread is blocked on an MVar operation, but in this case the `putMVar` can succeed whereas in the real GHC MVar impl such an exception will cause the `putMVar` to have no effect.

I think this shows that it's not actually possible to implement MVars using STM if one also cares about async exception behaviour.

Now arguably for the simulation this behaviour can't happen due to the scheduler (though perhaps can with POR), but then that argument would also suggest that this "default" implementation should be moved to the simulation instance and not be provided in `io-classes`.

Reply:
Ok, I've moved the MVarDefault from `io-classes` to `io-sim`.